File: src/main/java/sk/hudak/prco/service/impl/InternalTxServiceImpl.java
package sk.hudak.prco.service.impl;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sk.hudak.prco.api.ErrorType;
import sk.hudak.prco.api.EshopUuid;
import sk.hudak.prco.dto.ProductStatisticInfoDto;
import sk.hudak.prco.dto.ProductUpdateDataDto;
import sk.hudak.prco.dto.UnitData;
import sk.hudak.prco.dto.WatchDogAddDto;
import sk.hudak.prco.dto.WatchDogDto;
import sk.hudak.prco.dto.WatchDogNotifyUpdateDto;
import sk.hudak.prco.dto.error.ErrorCreateDto;
import sk.hudak.prco.dto.error.ErrorFindFilterDto;
import sk.hudak.prco.dto.error.ErrorListDto;
import sk.hudak.prco.dto.group.GroupCreateDto;
import sk.hudak.prco.dto.group.GroupFilterDto;
import sk.hudak.prco.dto.group.GroupIdNameDto;
import sk.hudak.prco.dto.group.GroupListDto;
import sk.hudak.prco.dto.group.GroupListExtendedDto;
import sk.hudak.prco.dto.group.GroupProductKeywordsCreateDto;
import sk.hudak.prco.dto.group.GroupProductKeywordsFullDto;
import sk.hudak.prco.dto.group.GroupUpdateDto;
import sk.hudak.prco.dto.internal.StatisticForUpdateForEshopDto;
import sk.hudak.prco.dto.newproduct.NewProductCreateDto;
import sk.hudak.prco.dto.newproduct.NewProductFilterUIDto;
import sk.hudak.prco.dto.newproduct.NewProductFullDto;
import sk.hudak.prco.dto.newproduct.NewProductInfoDetail;
import sk.hudak.prco.dto.notinteretedproduct.NotInterestedProductFindDto;
import sk.hudak.prco.dto.notinteretedproduct.NotInterestedProductFullDto;
import sk.hudak.prco.dto.product.ProductAddingToGroupDto;
import sk.hudak.prco.dto.product.ProductBestPriceInGroupDto;
import sk.hudak.prco.dto.product.ProductDetailInfo;
import sk.hudak.prco.dto.product.ProductFilterUIDto;
import sk.hudak.prco.dto.product.ProductFullDto;
import sk.hudak.prco.dto.product.ProductInActionDto;
import sk.hudak.prco.dto.product.ProductUnitDataDto;
import sk.hudak.prco.service.ErrorService;
import sk.hudak.prco.service.GroupProductKeywordsService;
import sk.hudak.prco.service.GroupService;
import sk.hudak.prco.service.InternalTxService;
import sk.hudak.prco.service.NewProductService;
import sk.hudak.prco.service.NotInterestedProductService;
import sk.hudak.prco.service.ProductCommonService;
import sk.hudak.prco.service.ProductService;
import sk.hudak.prco.service.WatchDogService;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Future;

/**
 * Transaction wrapper for internal services.
 */
@Slf4j
@Service
public class InternalTxServiceImpl implements InternalTxService {

    private final NewProductService newProductService;
    private final ProductCommonService productCommonService;
    private final ProductService productService;
    private final NotInterestedProductService notInterestedProductService;
    private final GroupService groupService;
    private final WatchDogService watchDogService;
    private final ErrorService errorService;
    private final GroupProductKeywordsService groupProductKeywordsService;

    public InternalTxServiceImpl(@Qualifier("newProductService") NewProductService newProductService,
                                 @Qualifier("productCommonService") ProductCommonService productCommonService,
                                 @Qualifier("productService") ProductService productService,
                                 @Qualifier("notInterestedProductService") NotInterestedProductService notInterestedProductService,
                                 @Qualifier("groupService") GroupService groupService,
                                 @Qualifier("watchDogService") WatchDogService watchDogService,
                                 @Qualifier("errorService") ErrorService errorService,
                                 @Qualifier("groupProductKeywordsService") GroupProductKeywordsService groupProductKeywordsService) {

        this.newProductService = newProductService;
        this.productCommonService = productCommonService;
        this.productService = productService;
        this.notInterestedProductService = notInterestedProductService;
        this.groupService = groupService;
        this.watchDogService = watchDogService;
        this.errorService = errorService;
        this.groupProductKeywordsService = groupProductKeywordsService;
    }

    @Override
    @Transactional(readOnly = true)
    public boolean existProductWithURL(String productURL) {
        return productCommonService.existProductWithURL(productURL);
    }

    @Override
    @Transactional
    public long createNewProduct(NewProductCreateDto newProductCreateDto) {
        return newProductService.createNewProduct(newProductCreateDto);
    }

    @Override
    @Transactional(readOnly = true)
    public NewProductFullDto getNewProduct(long newProductId) {
        return newProductService.getNewProduct(newProductId);
    }

    @Override
    @Transactional(readOnly = true)
    public NewProductInfoDetail findFirstInvalidNewProduct() {
        return newProductService.findFirstInvalidNewProduct();
    }

    @Override
    @Transactional(readOnly = true)
    public long getCountOfInvalidNewProduct() {
        return newProductService.getCountOfInvalidNewProduct();
    }

    @Override
    @Transactional
    public void repairInvalidUnitForNewProduct(Long newProductId,
                                               UnitData correctUnitData) {
        newProductService.repairInvalidUnitForNewProduct(newProductId, correctUnitData);
    }

    @Override
    @Transactional
    public void reprocessProductData(Long newProductId) {
        newProductService.reprocessProductData(newProductId);
    }

    @Override
    @Transactional
    public void confirmUnitDataForNewProducts(Long... newProductIds) {
        newProductService.confirmUnitDataForNewProducts(newProductIds);
    }

    @Override
    @Transactional
    public long fixAutomaticallyProductUnitData(int maxCountOfInvalid) {
        return newProductService.fixAutomaticallyProductUnitData(maxCountOfInvalid);
    }

    @Override
    @Transactional(readOnly = true)
    public List<NewProductFullDto> findNewProducts(NewProductFilterUIDto filter) {
        return newProductService.findNewProducts(filter);
    }

    @Override
    @Transactional
    public void markNewProductAsInterested(Long... newProductIds) {
        productCommonService.markNewProductAsInterested(newProductIds);
    }

    @Override
    @Transactional
    public void markNewProductAsNotInterested(Long... newProductIds) {
        productCommonService.markNewProductAsNotInterested(newProductIds);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<ProductDetailInfo> getProductForUpdate(EshopUuid eshopUuid, int olderThanInHours) {
        return productService.getProductForUpdate(eshopUuid, olderThanInHours);
    }

    @Override
    @Transactional
    public void updateProduct(ProductUpdateDataDto productUpdateDataDto) {
        productService.updateProduct(productUpdateDataDto);
    }

    @Override
    @Transactional
    public void updateProductUnitData(ProductUnitDataDto productUnitDataDto) {
        newProductService.updateProductUnitData(productUnitDataDto);
    }

    @Override
    @Transactional
    public void deleteNewProducts(Long... newProductIds) {
        newProductService.deleteNewProducts(newProductIds);
    }

    @Override
    @Transactional(readOnly = true)
    public long getCountOfAllNewProducts() {
        return newProductService.getCountOfAllNewProducts();
    }

    @Override
    @Transactional
    public void deleteNotInterestedProducts(EshopUuid eshopUuid) {
        notInterestedProductService.deleteNotInterestedProducts(eshopUuid);
    }

    @Override
    @Transactional
    public void markProductAsUnavailable(Long productId) {
        productService.markProductAsUnavailable(productId);
    }

    @Override
    @Transactional
    public void resetUpdateDateProduct(Long productId) {
        productService.resetUpdateDateProduct(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductBestPriceInGroupDto> findProductsBestPriceInGroupDto(EshopUuid eshopUuid) {
        return productService.findProductsBestPriceInGroupDto(eshopUuid);
    }

    @Override
    @Transactional(readOnly = true)
    public StatisticForUpdateForEshopDto getStatisticForUpdateForEshop(EshopUuid eshopUuid, int olderThanInHours) {
        return productService.getStatisticForUpdateForEshop(eshopUuid, olderThanInHours);
    }

    @Override
    @Transactional
    public void markProductAsNotInterested(Long productId) {
        productService.markProductAsNotInterested(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<NewProductFullDto> findNewProductsForExport() {
        return newProductService.findNewProductsForExport();
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductFullDto> findProductsForExport() {
        return productService.findProductsForExport();
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductFullDto> findProducts(ProductFilterUIDto filter) {
        return productService.findProducts(filter);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductInActionDto> findProductsInAction(EshopUuid eshopUuid) {
        return productService.findProductsInAction(eshopUuid);
    }

    @Override
    @Transactional

    public void removeProduct(Long productId) {
        productService.removeProduct(productId);
    }

    @Override
    @Transactional
    public void removeProductByUrl(String productUrl) {
        productService.removeProductByUrl(productUrl);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductFullDto> findProductsInGroup(Long groupId, boolean withPriceOnly, EshopUuid... eshopsToSkip) {
        return productService.findProductsInGroup(groupId, withPriceOnly, eshopsToSkip);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductFullDto> findProductsNotInAnyGroup() {
        return productService.findProductsNotInAnyGroup();
    }

    @Override
    @Transactional(readOnly = true)
    public ProductAddingToGroupDto getProduct(Long productId) {
        return productService.getProduct(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public boolean existProductWithUrl(String productURL) {
        return productService.existProductWithUrl(productURL);
    }

    @Override
    @Transactional
    public void resetUpdateDateForAllProductsInEshop(EshopUuid eshopUuid) {
        productService.resetUpdateDateForAllProductsInEshop(eshopUuid);
    }

    @Override
    @Transactional
    public void updateProductCommonPrice(Long productId, BigDecimal newCommonPrice) {
        productService.updateProductCommonPrice(productId, newCommonPrice);
    }

    @Override
    @Transactional(readOnly = true)
    public EshopUuid getEshopForProductId(Long productId) {
        return productService.getEshopForProductId(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public ProductDetailInfo getProductForUpdate(Long productId) {
        return productService.getProductForUpdate(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<NotInterestedProductFullDto> findNotInterestedProducts(NotInterestedProductFindDto findDto) {
        return productCommonService.findNotInterestedProducts(findDto);
    }

    @Override
    @Transactional
    public long importNewProducts(List<NewProductFullDto> newProductList) {
        return productCommonService.importNewProducts(newProductList);
    }

    @Override
    @Transactional
    public long importProducts(List<ProductFullDto> productList) {
        return productCommonService.importProducts(productList);
    }

    @Override
    @Transactional
    public long importNotInterestedProducts(List<NotInterestedProductFullDto> notInterestedProductList) {
        return productCommonService.importNotInterestedProducts(notInterestedProductList);
    }

    @Override
    @Transactional(readOnly = true)
    public ProductStatisticInfoDto getStatisticsOfProducts() {
        return productCommonService.getStatisticsOfProducts();
    }

    // group

    @Override
    @Transactional
    public Long createGroup(GroupCreateDto createDto) {
        return groupService.createGroup(createDto);
    }

    @Override
    @Transactional
    public void updateGroup(GroupUpdateDto updateDto) {
        groupService.updateGroup(updateDto);
    }

    @Override
    @Transactional
    public void addProductsToGroup(Long groupId, Long... productId) {
        groupService.addProductsToGroup(groupId, productId);
    }

    @Override
    @Transactional
    public void removeProductsFromGroup(Long groupId, Long... productIds) {
        groupService.removeProductsFromGroup(groupId, productIds);
    }

    @Override
    @Transactional(readOnly = true)
    public List<GroupListDto> getGroupsWithoutProduct(Long productId) {
        return groupService.getGroupsWithoutProduct(productId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<GroupListDto> findGroups(GroupFilterDto groupFilterDto) {
        return groupService.findGroups(groupFilterDto);
    }

    @Override
    @Transactional(readOnly = true)
    public List<GroupListExtendedDto> findAllGroupExtended() {
        return groupService.findAllGroupExtended();
    }

    @Override
    @Transactional(readOnly = true)
    public GroupIdNameDto getGroupById(Long groupId) {
        return groupService.getGroupById(groupId);
    }

    @Override
    @Transactional
    public Long addNewProductToWatch(WatchDogAddDto addDto) {
        return watchDogService.addNewProductToWatch(addDto);
    }

    @Override
    @Transactional(readOnly = true)
    public Map<EshopUuid, List<WatchDogDto>> findProductsForWatchDog() {
        return watchDogService.findProductsForWatchDog();
    }

    @Override
    public void notifyByEmail(List<WatchDogNotifyUpdateDto> toBeNotified) {
        // nemusi bezat v tranzakcii
        watchDogService.notifyByEmail(toBeNotified);
    }


    @Override
    @Transactional
    public Long createError(ErrorCreateDto createDto) {
        return errorService.createError(createDto);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ErrorListDto> findAll() {
        return errorService.findAll();
    }

    @Override
    @Transactional(readOnly = true)
    public List<ErrorListDto> findErrorByMaxCount(int limit, ErrorType errorType) {
        return errorService.findErrorByMaxCount(limit, errorType);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ErrorListDto> findErrorsByFilter(ErrorFindFilterDto findDto) {
        return errorService.findErrorsByFilter(findDto);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ErrorListDto> findErrorsByTypes(ErrorType... errorTypes) {
        return errorService.findErrorsByTypes(errorTypes);
    }

    @Override
    @Transactional(readOnly = true)
    public Map<ErrorType, Long> getStatisticForErrors() {
        return errorService.getStatisticForErrors();
    }

    @Override
    @Transactional
    public Future<Void> startErrorCleanUp() {
        return errorService.startErrorCleanUp();
    }

    @Override
    @Transactional
    public void deleteNotInterestedProducts(Long... notInterestedProductIds) {
        notInterestedProductService.deleteNotInterestedProducts(notInterestedProductIds);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ProductFullDto> findDuplicityProductsByNameAndPriceInEshop(EshopUuid eshopUuid) {
        return productService.findDuplicityProductsByNameAndPriceInEshop(eshopUuid);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Long> getProductWithUrl(String productUrl, Long productIdToIgnore) {
        return productService.getProductWithUrl(productUrl, productIdToIgnore);
    }

    @Override
    @Transactional
    public Long createGroupProductKeywords(GroupProductKeywordsCreateDto groupProductKeywordsCreateDto) {
        return groupProductKeywordsService.createGroupProductKeywords(groupProductKeywordsCreateDto);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<GroupProductKeywordsFullDto> getGroupProductKeywordsByGroupId(Long groupId) {
        return groupProductKeywordsService.getGroupProductKeywordsByGroupId(groupId);
    }

    @Override
    @Transactional
    public void removeAllKeywordForGroupId(Long groupId) {
        groupProductKeywordsService.removeAllKeywordForGroupId(groupId);
    }

    // tests

    @Override
    @Transactional
    public void test() {
        throw new UnsupportedOperationException();
    }
}

File: src/main/java/sk/hudak/prco/service/impl/InternalTxServiceImpl.kt
package sk.hudak.prco.service.impl

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import sk.hudak.prco.api.ErrorType
import sk.hudak.prco.api.EshopUuid
import sk.hudak.prco.dto.*
import sk.hudak.prco.dto.product.*
import sk.hudak.prco.service.*
import java.math.BigDecimal
import java.util.*
import java.util.concurrent.Future

/**
 * Transaction wrapper for internal services.
 */
@Service
open class InternalTxServiceImpl(@param:Qualifier("newProductService") private val newProductService: NewProductService,
                                 @param:Qualifier("productCommonService") private val productCommonService: ProductCommonService,
                                 @param:Qualifier("productService") private val productService: ProductService,
                                 @param:Qualifier("notInterestedProductService") private val notInterestedProductService: NotInterestedProductService,
                                 @param:Qualifier("groupService") private val groupService: GroupService,
                                 @param:Qualifier("watchDogService") private val watchDogService: WatchDogService,
                                 @param:Qualifier("errorService") private val errorService: ErrorService,
                                 @param:Qualifier("groupProductKeywordsService") private val groupProductKeywordsService: GroupProductKeywordsService,
                                 @param:Qualifier("searchKeywordService") private val searchKeywordService: SearchKeywordService)
    : InternalTxService {

    @Transactional
    override fun createSearchKeyword(createDto: SearchKeywordCreateDto): Long =
            searchKeywordService.createSearchKeyword(createDto)

    @Transactional
    override fun updateSearchKeyword(updateDto: SearchKeywordUdateDto) {
        searchKeywordService.updateSearchKeyword(updateDto)
    }

    @Transactional(readOnly = true)
    override fun findAllSearchKeyword(): List<SearchKeywordListDto> =
        searchKeywordService.findAllSearchKeyword()

    @Transactional
    override fun updateProductUrl(productId: Long, newProductUrl: String) {
        productService.updateProductUrl(productId, newProductUrl)
    }

    @Transactional(readOnly = true)
    override fun getProductForUpdateByUrl(productUrl: String): ProductDetailInfo? =
            productService.getProductForUpdateByUrl(productUrl)


    @Transactional(readOnly = true)
    override fun findProductForUpdateInGroup(groupId: Long): Map<EshopUuid, List<Long>> =
            productService.findProductForUpdateInGroup(groupId)

    @Transactional(readOnly = true)
    override fun findProductsForUpdateWhichAreNotInAnyGroup(): Map<EshopUuid, List<Long>> =
            productService.findProductsForUpdateWhichAreNotInAnyGroup()

    @Transactional
    override fun removeWatchDog(eshopUuid: EshopUuid, maxCountToDelete: Long): Int =
            watchDogService.removeWatchDog(eshopUuid, maxCountToDelete)

    @Transactional
    override fun removeErrorsByCount(eshopUuid: EshopUuid, maxCountToDelete: Long): Int =
            errorService.removeErrorsByCount(eshopUuid, maxCountToDelete)


    @Transactional
    override fun removeProductsByCount(eshopUuid: EshopUuid, maxCountToDelete: Long): Int =
            productService.removeProductsByCount(eshopUuid, maxCountToDelete)


    @Transactional
    override fun removeNewProductsByCount(eshopUuid: EshopUuid, maxCountToDelete: Long): Int =
            newProductService.removeNewProductsByCount(eshopUuid, maxCountToDelete)


    @Transactional
    override fun removeNotInterestedProductsByCount(eshopUuid: EshopUuid, maxCountToDelete: Long): Int {
        return notInterestedProductService.removeNotInterestedProductsByCount(eshopUuid, maxCountToDelete)
    }

    override val countOfInvalidNewProduct: Long
        @Transactional(readOnly = true)
        get() = newProductService.countOfInvalidNewProduct

    override val countOfAllNewProducts: Long
        @Transactional(readOnly = true)
        get() = newProductService.countOfAllNewProducts

    override val statisticsOfProducts: ProductStatisticInfoDto
        @Transactional(readOnly = true)
        get() = productCommonService.statisticsOfProducts

    override val statisticForErrors: Map<ErrorType, Long>
        @Transactional(readOnly = true)
        get() = errorService.statisticForErrors

    @Transactional(readOnly = true)
    override fun existProductWithURL(productURL: String): Boolean {
        return productCommonService.existProductWithURL(productURL)
    }

    @Transactional
    override fun createNewProduct(newProductCreateDto: NewProductCreateDto): Long {
        return newProductService.createNewProduct(newProductCreateDto)
    }

    @Transactional(readOnly = true)
    override fun getNewProduct(newProductId: Long): NewProductFullDto {
        return newProductService.getNewProduct(newProductId)
    }

    @Transactional(readOnly = true)
    override fun findFirstInvalidNewProduct(): NewProductInfoDetail? {
        return newProductService.findFirstInvalidNewProduct()
    }

    @Transactional
    override fun repairInvalidUnitForNewProduct(newProductId: Long?,
                                                correctUnitData: UnitData) {
        newProductService.repairInvalidUnitForNewProduct(newProductId, correctUnitData)
    }

    @Transactional
    override fun reprocessProductData(newProductId: Long?) {
        newProductService.reprocessProductData(newProductId)
    }

    @Transactional
    override fun confirmUnitDataForNewProducts(newProductIds: LongArray) {
        newProductService.confirmUnitDataForNewProducts(newProductIds)
    }

    @Transactional
    override fun fixAutomaticallyProductUnitData(maxCountOfInvalid: Int): Long {
        return newProductService.fixAutomaticallyProductUnitData(maxCountOfInvalid)
    }

    @Transactional(readOnly = true)
    override fun findNewProducts(filter: NewProductFilterUIDto): List<NewProductFullDto> {
        return newProductService.findNewProducts(filter)
    }

    @Transactional
    override fun markNewProductAsInterested(vararg newProductIds: Long) {
        productCommonService.markNewProductAsInterested(*newProductIds)
    }

    @Transactional
    override fun markNewProductAsNotInterested(vararg newProductIds: Long) {
        productCommonService.markNewProductAsNotInterested(*newProductIds)
    }

    @Transactional(readOnly = true)
    override fun findProductForUpdate(eshopUuid: EshopUuid, olderThanInHours: Int): ProductDetailInfo? {
        return productService.findProductForUpdate(eshopUuid, olderThanInHours)
    }

    @Transactional
    override fun updateProduct(productUpdateDataDto: ProductUpdateDataDto) {
        productService.updateProduct(productUpdateDataDto)
    }

    @Transactional
    override fun updateProductUnitData(productUnitDataDto: ProductUnitDataDto) {
        newProductService.updateProductUnitData(productUnitDataDto)
    }

    @Transactional
    override fun deleteNewProducts(newProductIds: Array<Long>) {
        newProductService.deleteNewProducts(newProductIds)
    }

    @Transactional
    override fun deleteNotInterestedProducts(eshopUuid: EshopUuid) {
        notInterestedProductService.deleteNotInterestedProducts(eshopUuid)
    }

    @Transactional
    override fun markProductAsUnavailable(productId: Long) {
        productService.markProductAsUnavailable(productId)
    }

    @Transactional
    override fun resetUpdateDateProduct(productId: Long) {
        productService.resetUpdateDateProduct(productId)
    }

    @Transactional(readOnly = true)
    override fun findProductsBestPriceInGroupDto(eshopUuid: EshopUuid): List<ProductBestPriceInGroupDto> {
        return productService.findProductsBestPriceInGroupDto(eshopUuid)
    }

    @Transactional(readOnly = true)
    override fun getStatisticForUpdateForEshop(eshopUuid: EshopUuid, olderThanInHours: Int): StatisticForUpdateForEshopDto {
        return productService.getStatisticForUpdateForEshop(eshopUuid, olderThanInHours)
    }

    @Transactional
    override fun markProductAsNotInterested(productId: Long) {
        productService.markProductAsNotInterested(productId)
    }

    @Transactional(readOnly = true)
    override fun findNewProductsForExport(): List<NewProductFullDto> {
        return newProductService.findNewProductsForExport()
    }

    @Transactional(readOnly = true)
    override fun findProductsForExport(): List<ProductFullDto> {
        return productService.findProductsForExport()
    }

    @Transactional(readOnly = true)
    override fun findProducts(filter: ProductFilterUIDto): List<ProductFullDto> {
        return productService.findProducts(filter)
    }

    @Transactional(readOnly = true)
    override fun findProductsInAction(eshopUuid: EshopUuid): List<ProductInActionDto> {
        return productService.findProductsInAction(eshopUuid)
    }

    @Transactional
    override fun removeProduct(productId: Long) {
        productService.removeProduct(productId)
    }

    @Transactional
    override fun removeProductByUrl(productUrl: String) {
        productService.removeProductByUrl(productUrl)
    }

    @Transactional(readOnly = true)
    override fun findProductsInGroup(groupId: Long, withPriceOnly: Boolean, vararg eshopsToSkip: EshopUuid): List<ProductFullDto> {
        return productService.findProductsInGroup(groupId, withPriceOnly, *eshopsToSkip)
    }

    @Transactional(readOnly = true)
    override fun findProductsNotInAnyGroup(): List<ProductFullDto> {
        return productService.findProductsNotInAnyGroup()
    }

    @Transactional(readOnly = true)
    override fun getProduct(productId: Long?): ProductAddingToGroupDto {
        return productService.getProduct(productId)
    }

    @Transactional(readOnly = true)
    override fun existProductWithUrl(productURL: String): Boolean {
        return productService.existProductWithUrl(productURL)
    }

    @Transactional
    override fun resetUpdateDateForAllProductsInEshop(eshopUuid: EshopUuid) {
        productService.resetUpdateDateForAllProductsInEshop(eshopUuid)
    }

    @Transactional
    override fun updateProductCommonPrice(productId: Long, newCommonPrice: BigDecimal) {
        productService.updateProductCommonPrice(productId, newCommonPrice)
    }

    @Transactional(readOnly = true)
    override fun getEshopForProductId(productId: Long): EshopUuid {
        return productService.getEshopForProductId(productId)
    }

    @Transactional(readOnly = true)
    override fun findProductForUpdate(productId: Long): ProductDetailInfo {
        return productService.findProductForUpdate(productId)
    }

    @Transactional(readOnly = true)
    override fun findNotInterestedProducts(findDto: NotInterestedProductFindDto): List<NotInterestedProductFullDto> {
        return notInterestedProductService.findNotInterestedProducts(findDto)
    }

    @Transactional
    override fun importNewProducts(newProductList: List<NewProductFullDto>): Long {
        return productCommonService.importNewProducts(newProductList)
    }

    @Transactional
    override fun importProducts(productList: List<ProductFullDto>): Long {
        return productCommonService.importProducts(productList)
    }

    @Transactional
    override fun importNotInterestedProducts(notInterestedProductList: List<NotInterestedProductFullDto>): Long {
        return productCommonService.importNotInterestedProducts(notInterestedProductList)
    }

    // group

    @Transactional
    override fun createGroup(createDto: GroupCreateDto): Long {
        return groupService.createGroup(createDto)
    }

    @Transactional
    override fun updateGroup(updateDto: GroupUpdateDto) {
        groupService.updateGroup(updateDto)
    }

    @Transactional
    override fun addProductsToGroup(groupId: Long, vararg productId: Long) {
        groupService.addProductsToGroup(groupId, *productId)
    }

    @Transactional
    override fun removeProductsFromGroup(groupId: Long?, vararg productIds: Long) {
        groupService.removeProductsFromGroup(groupId, *productIds)
    }

    @Transactional(readOnly = true)
    override fun getGroupsWithoutProduct(productId: Long?): List<GroupListDto> {
        return groupService.getGroupsWithoutProduct(productId)
    }

    @Transactional(readOnly = true)
    override fun findGroups(groupFilterDto: GroupFilterDto): List<GroupListDto> {
        return groupService.findGroups(groupFilterDto)
    }

    @Transactional(readOnly = true)
    override fun findAllGroupExtended(): List<GroupListExtendedDto> {
        return groupService.findAllGroupExtended()
    }

    @Transactional(readOnly = true)
    override fun getGroupById(groupId: Long?): GroupIdNameDto {
        return groupService.getGroupById(groupId)
    }

    @Transactional
    override fun addNewProductToWatch(addDto: WatchDogAddCustomDto): Long? {
        return watchDogService.addNewProductToWatch(addDto)
    }

    @Transactional(readOnly = true)
    override fun findProductsForWatchDog(): Map<EshopUuid, List<WatchDogDto>> {
        return watchDogService.findProductsForWatchDog()
    }

    override fun notifyByEmail(toBeNotified: List<WatchDogNotifyUpdateDto>) {
        // nemusi bezat v tranzakcii
        watchDogService.notifyByEmail(toBeNotified)
    }


    @Transactional
    override fun createError(createDto: ErrorCreateDto): Long? {
        return errorService.createError(createDto)
    }

    @Transactional(readOnly = true)
    override fun findAll(): List<ErrorListDto> {
        return errorService.findAll()
    }

    @Transactional(readOnly = true)
    override fun findErrorByMaxCount(limit: Int, errorType: ErrorType): List<ErrorListDto> {
        return errorService.findErrorByMaxCount(limit, errorType)
    }

    @Transactional(readOnly = true)
    override fun findErrorsByFilter(findDto: ErrorFindFilterDto): List<ErrorListDto> {
        return errorService.findErrorsByFilter(findDto)
    }

    @Transactional(readOnly = true)
    override fun findErrorsByTypes(vararg errorTypes: ErrorType): List<ErrorListDto> {
        return errorService.findErrorsByTypes(*errorTypes)
    }

    @Transactional
    override fun startErrorCleanUp(): Future<Void>? {
        return errorService.startErrorCleanUp()
    }

    @Transactional
    override fun deleteNotInterestedProducts(vararg notInterestedProductIds: Long) {
        notInterestedProductService.deleteNotInterestedProducts(*notInterestedProductIds)
    }

    @Transactional(readOnly = true)
    override fun findDuplicityProductsByNameAndPriceInEshop(eshopUuid: EshopUuid): List<ProductFullDto> {
        return productService.findDuplicityProductsByNameAndPriceInEshop(eshopUuid)
    }

    @Transactional(readOnly = true)
    override fun getProductWithUrl(productUrl: String, productIdToIgnore: Long?): Optional<Long> {
        return productService.getProductWithUrl(productUrl, productIdToIgnore)
    }

    @Transactional
    override fun createGroupProductKeywords(groupProductKeywordsCreateDto: GroupProductKeywordsCreateDto): Long? {
        return groupProductKeywordsService.createGroupProductKeywords(groupProductKeywordsCreateDto)
    }

    @Transactional(readOnly = true)
    override fun getGroupProductKeywordsByGroupId(groupId: Long?): Optional<GroupProductKeywordsFullDto> {
        return groupProductKeywordsService.getGroupProductKeywordsByGroupId(groupId)
    }

    @Transactional
    override fun removeAllKeywordForGroupId(groupId: Long?) {
        groupProductKeywordsService.removeAllKeywordForGroupId(groupId)
    }

    // tests

    @Transactional
    override fun test() {
        throw UnsupportedOperationException()
    }
}
